libp2p/NodeTable.h

```diff
@@ -556,18 +576,75 @@ struct Neighbours: DiscoveryDatagram
         _s.appendList(neighbours.size());
         for (auto const& n: neighbours)
             n.streamRLP(_s);
-        _s << ts;
+        _s << *timestamp;
     }
     void interpretRLP(bytesConstRef _bytes) override
     {
         RLP r(_bytes, RLP::AllowNonCanon|RLP::ThrowOnFail);
         for (auto const& n: r[0])
             neighbours.emplace_back(n);
-        ts = r[1].toInt<uint32_t>();
+        timestamp = r[1].toInt<uint32_t>();
     }
 
     std::string typeName() const override { return "Neighbours"; }
 };
 
+struct ENRRequest : DiscoveryDatagram
+{
+    // Constructor for outgoing packets
+    ENRRequest(bi::udp::endpoint _to) : DiscoveryDatagram{_to} {}
+    // Constructor for incoming packets
+    ENRRequest(bi::udp::endpoint const& _from, NodeID const& _fromid, h256 const& _echo)
+      : DiscoveryDatagram{_from, _fromid, _echo}
+    {}
+
+    static uint8_t const type = 5;
+    uint8_t packetType() const override { return type; }
+
+    void streamRLP(RLPStream& _s) const override
+    {
+        _s.appendList(1);
+        _s << *timestamp;
+    }
+    void interpretRLP(bytesConstRef _bytes) override
+    {
+        RLP r(_bytes, RLP::AllowNonCanon | RLP::ThrowOnFail);
+        timestamp = r[0].toInt<uint32_t>();
+    }
+
+    std::string typeName() const override { return "ENRRequest"; }
+};
+
+struct ENRResponse : DiscoveryDatagram
+{
+    // Constructor for outgoing packets
+    ENRResponse(bi::udp::endpoint const& _dest, ENR const& _enr)
+      : DiscoveryDatagram{_dest}, enr{_enr}
+    {}
+    // Constructor for incoming packets
+    ENRResponse(bi::udp::endpoint const& _from, NodeID const& _fromid, h256 const& _echo)
+      : DiscoveryDatagram{_from, _fromid, _echo}
+    {}
+
+    static uint8_t const type = 6;
+    uint8_t packetType() const override { return type; }
+
+    boost::optional<ENR> enr;
```

Comment:
This is only because `ENR` doesn't have a default constructor (and I'd prefer to leave it like this - empty ENR wouldn't be valid), and it is created not in constructor, but in `interpretRLP` method.

Actually I think I'll change it to `unique_ptr`, it would better express the intent.